containers/src/Data/IntSet/Internal.hs

```diff
@@ -1724,45 +1720,53 @@ revWord x1 = case ((x1 `shiftRL` 1) .&. 0x5555555555555555) .|. ((x1 .&. 0x55555
                        x6 -> ( x6 `shiftRL` 32             ) .|. ( x6               `shiftLL` 32);
 #endif
 
-foldlBits prefix f z bitmap = go bitmap z
-  where go 0 acc = acc
-        go bm acc = go (bm `xor` bitmask) ((f acc) $! (prefix+bi))
-          where
-            !bitmask = lowestBitMask bm
-            !bi = countTrailingZeros bitmask
-
-foldl'Bits prefix f z bitmap = go bitmap z
-  where go 0 acc = acc
-        go bm !acc = go (bm `xor` bitmask) ((f acc) $! (prefix+bi))
-          where !bitmask = lowestBitMask bm
-                !bi = countTrailingZeros bitmask
-
-foldrBits prefix f z bitmap = go (revWord bitmap) z
-  where go 0 acc = acc
-        go bm acc = go (bm `xor` bitmask) ((f $! (prefix+(WORD_SIZE_IN_BITS-1)-bi)) acc)
-          where !bitmask = lowestBitMask bm
-                !bi = countTrailingZeros bitmask
-
-
-foldr'Bits prefix f z bitmap = go (revWord bitmap) z
-  where go 0 acc = acc
-        go bm !acc = go (bm `xor` bitmask) ((f $! (prefix+(WORD_SIZE_IN_BITS-1)-bi)) acc)
-          where !bitmask = lowestBitMask bm
-                !bi = countTrailingZeros bitmask
-
-foldMapBits prefix f bitmap = go (prefix + bi0) (bitmap `xor` bitmask0)
+foldlBits prefix f z0 bitmap = go z0 $! revWord bitmap
+  where
+    -- Note: We pass the z as a static argument because it helps GHC with demand
+    -- analysis. See GHC #25578 for details.
+    go z !bm = f (if bm' == 0 then z else go z bm') (prefix .|. bi)
```

Comment:
Does GHC calculate the second argument eagerly when `f` is unknown? Should it?

Reply:
> Does GHC calculate the second argument eagerly when `f` is unknown?

GHC decides to do that here, yes. The STG looks like

```
let {
  Rec {
  $wgo8_s4BNB [InlPrag=[2], Occ=LoopBreaker, Dmd=SC(S,C(1,L))]
    :: a_a4wJ1 -> Word# -> a_a4wJ1
  [LclId, Arity=2, Str=<ML><L>, Unf=OtherCon []] =
      {$wgo8_s4BNB, f_s4BNo, bx_s4BNz} \r [z1_s4BNC ww_s4BND]
          case ctz# [ww_s4BND] of sat_s4BNH {
          __DEFAULT ->
          case word2Int# [sat_s4BNH] of sat_s4BNI {
          __DEFAULT ->
          case -# [63# sat_s4BNI] of sat_s4BNJ {
          __DEFAULT ->
          case orI# [bx_s4BNz sat_s4BNJ] of sat_s4BNK {
          __DEFAULT ->
          let {
            sat_s4BNL [Occ=Once1] :: Int
            [LclId] =
                I#! [sat_s4BNK]; } in
          let {
            sat_s4BNG [Occ=Once1] :: a_a4wJ1
            [LclId] =
                {z1_s4BNC, $wgo8_s4BNB, ww_s4BND} \u []
                    case minusWord# [ww_s4BND 1##] of sat_s4BNE {
                    __DEFAULT ->
                    case and# [ww_s4BND sat_s4BNE] of wild1_s4BNF {
                      __DEFAULT -> $wgo8_s4BNB z1_s4BNC wild1_s4BNF;
                      0## -> z1_s4BNC;
                    };
                    };
          } in  f_s4BNo sat_s4BNG sat_s4BNL;
          };
          };
          };
          };
  end Rec }
} in
...
```

> Should it?

I don't see any issue with leaving this up to GHC. Do you?  
Compared the forcing it always, the (perhaps small) benefit is that GHC can optimize for some known `f` to drop this calculation if it isn't used.

Reply:
The only time that could happen would be if the result only depends on the size of the set. I believe the general case can be handled by a function something like this:

```haskell
sc :: IntSet -> (Int -> StateT s (Either r) ()) -> s -> r
```

The function argument should have the property that `f m *> f n = f (m + n)` when `m` and `n` are positive.

Reply:
> The only time that could happen would be if the result only depends on the size of the set.

No, why would that be? Let's say we want to sum every 10th element in the set.

```hs
sum10 :: IntSet -> Int
sum10 s = foldr f z s (0 :: Int) 0
  where
    f x k !i !acc = k (if i == 9 then 0 else i+1) (if i == 0 then acc+x else acc)
    z !_ acc = acc
```

This doesn't care about 90% of the elements, and I expect GHC to optimize this to skip all the unnecessary `prefix .|. ctz bm`.

Reply:
I checked the Core for `sum10` and it turned to be identical when forcing and not forcing the element. GHC did the reasonable thing and moved the `prefix .|. ctz bm` to the branch where `i == 0`.

> Should it?

So the answer to this seems to me to be "it doesn't matter". But if you would prefer to force the element I can make the change.

Reply:
I think it's probably better to force it, so no one else needs to wonder. But I don't feel super strongly about it.